### PR TITLE
Websocket: support android api 21

### PR DESCRIPTION
**Pull request by luca992**
Just had to replace use of java.util.Base64 with io.ktor.util.Base64. And use `SecureRandom()` when `SecureRandom.getInstanceStrong()` isn't available. `SecureRandom.getInstanceStrong()` is only available on Android API >= 26.

Tested with Android 5.0 (API level 21) and it works great! May work on even lower versions.... haven't tried though.

### CODE REVIEW

File: ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/TLSConfigBuilder.kt [hunk 1]
```diff
@@ -48,11 +48,20 @@ class TLSConfigBuilder {
...
-        certificates, trustManager as? X509TrustManager ?: findTrustManager(),
-        cipherSuites, serverName
-    )
+    fun build(): TLSConfig {
+        val secureRandomInstance = try {
+            SecureRandom.getInstanceStrong()
+        } catch (e: NoSuchMethodError) {
+            //SecureRandom.getInstanceStrong() requires JVM 1.8 and for Android api level 26
+            //if getInstanceStrong() is not available use non-blocking SecureRandom() to get instance
+            SecureRandom()
```
Comment (JLLeitschuh) [inline ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/TLSConfigBuilder.kt]: What are the security guarantees of `SecureRandom.getInstanceStrong()` that `SecureRandom()` won't give users? Are these equivalent?

Can you include a link to some sort of official Android documentation about how to generate secure random numbers in versions of android prior to 26?

Comment (luca992) [reply ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/TLSConfigBuilder.kt]: https://tersesystems.com/blog/2015/12/17/the-right-way-to-use-securerandom/
According to this article SecureRandom.getInstanceStrong() blocks the thread if there isn't enough randomness available.. And that pretty much won't happen unless you are running on a `container based solution like Docker or CoreOS`. Other than that they are equally secure... 

https://stackoverflow.com/questions/27622625/securerandom-with-nativeprng-vs-sha1prng
[The top answer](https://stackoverflow.com/a/27638413/1363742)  advocates that specifying a provider may hurt interoperability and [links](https://android-developers.googleblog.com/2016/06/security-crypto-provider-deprecated-in.html) to an android dev post where they say `SecureRandom()` should be used

Also,
https://www.veracode.com/blog/research/cryptographically-secure-pseudo-random-number-generator-csprng#tldr

Suggests: 
> Note: This recommendation has the additional advantage of keeping code portable across operating systems, and will provide a secure randomizer if self-seeded. If you want complete assurance of randomness for a given operating system, I would suggest explicitly using the "Windows-PRNG" algorithm for Windows environments (using the getInstance method) and "NativePRNG" for Unix-like environments. Note that these options carry the downside of making code not easily portable. This is explained in detail later in this pos

Comment (JLLeitschuh) [reply ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/TLSConfigBuilder.kt]: Sounds good. Thanks!

**Comment by luca992** [comment]
Any chance this will be considered being pulled for 1.2.0 release? Android 21 support is important for my project

**Comment by cruddell001** [comment]
This should be considered for more than just websocket support.  I have an app that uses Basic Auth and won't run on an older device due to ktor's use of java.util.Base64.  I will need to change networking libraries unless/until this is fixed.

**Comment by luca992** [comment]
@cruddell001 I can look tommorrow. But I also use basic auth and I dont remeber running into your issue on Android 21. Which class is causing the error specifically?

**Comment by cruddell001** [comment]
> @cruddell001 I can look tommorrow. But I also use basic auth and I dont remeber running into your issue on Android 21. Which class is causing the error specifically?

This is the stack trace:

```06-21 09:08:49.532 5325-5396/com.example E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: com.example, PID: 5325
    java.lang.NoClassDefFoundError: Failed resolution of: Ljava/util/Base64;
        at io.ktor.util.Base64JvmKt.encodeBase64(Base64Jvm.kt:15)
        at io.ktor.client.features.auth.providers.BasicAuthProvider.constructBasicAuthValue(BasicAuthProvider.kt:64)
        at io.ktor.client.features.auth.providers.BasicAuthProvider.addRequestHeaders(BasicAuthProvider.kt:59)
        at io.ktor.client.features.auth.Auth$Feature$install$1.invokeSuspend(Auth.kt:55)
        at io.ktor.client.features.auth.Auth$Feature$install$1.invoke(Auth.kt)
        at io.ktor.client.features.HttpSend$Feature$install$1.invokeSuspend(HttpSend.kt:65)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:211)
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:168)
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63)
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:118)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "java.util.Base64" on path: DexPathList[[zip file "/data/app/com.example-1/base.apk"],nativeLibraryDirectories=[/data/app/com.example-1/lib/arm, /vendor/lib, /system/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at io.ktor.util.Base64JvmKt.encodeBase64(Base64Jvm.kt:15) 
        at io.ktor.client.features.auth.providers.BasicAuthProvider.constructBasicAuthValue(BasicAuthProvider.kt:64) 
        at io.ktor.client.features.auth.providers.BasicAuthProvider.addRequestHeaders(BasicAuthProvider.kt:59) 
        at io.ktor.client.features.auth.Auth$Feature$install$1.invokeSuspend(Auth.kt:55) 
        at io.ktor.client.features.auth.Auth$Feature$install$1.invoke(Auth.kt) 
        at io.ktor.client.features.HttpSend$Feature$install$1.invokeSuspend(HttpSend.kt:65) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:211) 
        at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:168) 
        at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:63) 
        at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:118) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46) 
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742) 
    	Suppressed: java.lang.ClassNotFoundException: java.util.Base64
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 17 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
```

**Comment by luca992** [comment]
@cruddell001 that is what this pr fixes though?

**Comment by cruddell001** [comment]
@luca992 - I believe so.  As far as I can tell, this class (Base64Jvm.kt) is the only class in ktor that uses java.util.base64, and this pr appears to remove that dependency.  Base64Jvm.kt is also the class that triggered the exception in my app, as indicated in the stack trace.

**Comment by luca992** [comment]
So is therr an issue with my pr then? Or is there another dependency issue
that you have found?

**Comment by cruddell001** [comment]
No issue - I was simply saying this needs to be higher priority to get merged in as it effects more than just websockets.  I've switched over to Retrofit for my needs because of this bug.

**Comment by cy6erGn0m** [comment]
Base64 issue is already fixed in master

**Comment by cruddell001** [comment]
@cy6erGn0m Thanks for fixing that.  When will the next release be available with your change?

**Comment by cy6erGn0m** [comment]
@cruddell001 I believe, something about two weeks. With 1.3.41.

**Comment by cruddell001** [comment]
Ok thanks!

**Comment by luca992** [comment]
fixed in 1.2.3-rc

**Pull request closed without merging**